packages/core/src/utils/proxyUtils.ts

```diff
@@ -0,0 +1,146 @@
+/**
+ * Utility to redirect remote sources via Proxy
+ *
+ * This utility is used to rewrite a request to be routed through
+ * the Proxy instead.
+ */
+
+/**
+ * Helper-Function to route remote requests through a local proxy.
+ *
+ * @example This rewrites a remote url like `https://via.placeholder.com/300.png/09f/fff` into a URI-Component-Encoded string like `/cors-proxy/https%3A%2F%2Fvia.placeholder.com%2F300.png%2F09f%2Ffff`
+ */
+export function viaProxy(url: string) {
+  if (!isProxyEnabled()) {
+    // Proxy is disabled, so we just pass as-is.
+    return url;
+  }
+
+  if (url.startsWith('/cors-proxy/')) {
+    // Already proxied, return as-is
+    return url;
+  }
+
+  // window.location.hostname is being passed here to ensure that
+  // this does not throw an Error for same-origin requests
+  // e.g. /some/image -> localhost:9000/some/image
+  const selfUrl = new URL(window.location.toString());
+  // inside a try-catch in case the URL cannot be understood
+  try {
+    const expandedUrl = new URL(url, selfUrl);
+    if (!expandedUrl.protocol.startsWith('http')) {
+      // this is probably some embedded image (e.g. image/png;base64).
+      // don't touch and pass as is
+      return url;
+    }
+    if (selfUrl.host === expandedUrl.host) {
+      // This is a request to a "local" resource.
+      // No need to rewrite
+      return url;
+    }
+
+    // Check if the host matches the Allow List.
+    // if not, no rewrite takes place.
+    // will fail in the Editor if the
+    // remote host does not accept anonymous
+    if (!isInsideAllowList(expandedUrl.host)) {
+      return url;
+    }
+  } catch (_) {
+    // in case of error just silently pass as-is
+    return url;
+  }
+
+  // Everything else is a "remote" resource and requires a rewrite.
+  return `/cors-proxy/${encodeURIComponent(url)}`;
+}
+
+function isInsideAllowList(host: string) {
+  const allowList = getAllowList();
+  if (allowList.length === 0) {
+    // Allow List defaults to allow all if empty
+    return true;
+  }
+  for (const entry of allowList) {
+    if (entry.toLowerCase().trim() === host) {
+      return true;
+    }
+  }
+  return false;
+}
+
+/**
+ * This tries to access import.meta.env.VITE_MC_PROXY_ENABLED
+ * which will either return 'true' of 'false' (as strings) if
+ * present, or be undefined if not run from a vite context or
+ * run without the MC Plugin
+ *
+ * If no value was set this will be false as the proxy only
+ * exists inside the Vite Server.
+ */
+export function isProxyEnabled() {
+  try {
+    const importedValue = import.meta.env['VITE_MC_PROXY_ENABLED'] ?? 'false';
+    return importedValue === 'true';
+  } catch (_) {
+    return false;
+  }
+}
```

Comment:
The `try/catch` here shouldn't be necessary:
```suggestion
export function isProxyEnabled() {
  if (import.meta.env) {
    return import.meta.env.VITE_MC_PROXY_ENABLED === 'true';
  } else {
    return false;
  }
}
```
`import.meta` is a js standard so we can always access it to check if `env` exists.
This will also result in a cleaner output after minification:

```ts
// with `try/catch`:
function or(){try{return"true"==="true"}catch{return!1}}

// with `if`:
function or(){return!0}
```

I also don't know why this bracket notation is necessary:
```ts
import.meta.env['VITE_MC_PROXY_ENABLED']
```
I used a dot notation and it works fine:
```ts
import.meta.env.VITE_MC_PROXY_ENABLED
```
Vite replaces it when building, tests in Vitest pass. Is there any reason for it?

Reply:
> Vite replaces it when building, tests in Vitest pass. Is there any reason for it?

Oh . None at all. Dot notation should work fine